### PR TITLE
Create Grocery Example for Metrics API

### PR DESCRIPTION
**Pull request by victlu**
Create Grocery example per Metrics API SIG.

### CODE REVIEW

File: examples/GroceryExample/.vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,30 @@
+{
+    // Use IntelliSense to learn about possible attributes.
```
Comment (reyang) [inline examples/GroceryExample/.vscode/launch.json:2]: minor: remove the `.vscode` folder?

File: examples/GroceryExample/readme.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+implementation will likely affect how the API is designed, but we will make our
+best judgement to tolerate the situation at this time.
+
+## Topics for discussions
+
+- Need some kind of concrete LabelSet() in API.  Access to LabelSetSdk() is
+unavailable from API side.
+
+- It is inconvenient to have to pass a default(SpanContext) when we don't care
+about spans. Need additional prototypes to make SpanContext optional.
```
Comment (cijothomas) [inline examples/GroceryExample/readme.md:16]: yes. maybe we can just make it default to Activity.Current, Baggage.Current.

File: examples/GroceryExample/readme.md [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+- We create instrument with CreateInt64Counter() but it returns a generic
+Counter&lt;long&gt;. Seems like it should return a Int64Counter instead.
+
+- It's not allowed to new MeterProvider().  Thus, the only way to access is via
+the Default property. We can probably simplify MeterProvider.Default.GetMeter()
+to simply MeterProvider.GetMeter().
+
+- Bound counters does not allow adding more labels.  Ideally, we would bind
```
Comment (cijothomas) [inline examples/GroceryExample/readme.md:25]: yes this is something we need to solve. The bound counter in the original spec did not allow this. So bounds instruments were useable only if you know the key/value for *all* the labels, not just a part of them.

File: examples/GroceryExample/MyLabelSet.cs [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+// </copyright>
+
+using System.Collections.Generic;
+using OpenTelemetry.Metrics;
+
+#pragma warning disable CS0618
+
+namespace GroceryExample
+{
+    public class MyLabelSet : LabelSet
```
Comment (cijothomas) [inline examples/GroceryExample/MyLabelSet.cs:24]: trying to understand why did you create `MyLabelSet`?.... Help me understand.

File: examples/GroceryExample/GroceryStore.cs [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+                totalPrice += item.qty * priceList[item.name];
+
+                // Record Metric
+
+                var labels = new MyLabelSet(
+                    new KeyValuePair<string, string>("Store", "Portland"),
+                    new KeyValuePair<string, string>("Customer", customer),
+                    new KeyValuePair<string, string>("Item", item.name));
+
+                this.itemCounter.Add(default(SpanContext), item.qty, labels);
```
Comment (cijothomas) [inline examples/GroceryExample/GroceryStore.cs:74]: We can do similar to https://github.com/open-telemetry/opentelemetry-dotnet/blob/metrics/examples/Console/TestPrometheusExporter.cs#L88

File: examples/GroceryExample/GroceryStore.cs [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+            // Record Metric
+
+            var labels2 = new MyLabelSet(
+                new KeyValuePair<string, string>("Store", "Portland"),
+                new KeyValuePair<string, string>("Customer", customer));
+
+            this.cashCounter.Add(default(SpanContext), totalPrice, labels2);
+
+            this.boundCashCounter.Add(default(SpanContext), totalPrice);
```
Comment (cijothomas) [inline examples/GroceryExample/GroceryStore.cs:85]: This is not required as per https://github.com/open-telemetry/oteps/pull/146/files ?

File: examples/GroceryExample/MyLabelSet.cs (hunk 4 above)
Comment (victlu) [reply examples/GroceryExample/MyLabelSet.cs:24]: LabelSet is an abstract class in API.  And I don't want to take a dependency on SDK's LabelSetSdk.

Comment (cijothomas) [reply examples/GroceryExample/MyLabelSet.cs]: I see. You shouldn't need to create own LabelSet.. Meter exposes methods to create LabelSet. https://github.com/open-telemetry/opentelemetry-dotnet/blob/ab278ceccbf1e11a3829cad85a85d8a81ebb8539/src/OpenTelemetry.Api/Metrics/Meter.cs#L83

(you can ignore this comment as this code will anyway change a lot. but just sharing what was the original attempt)

Comment (victlu) [reply examples/GroceryExample/MyLabelSet.cs]: I see you have a Meter.GetLabelSet() virtual method.  I'll try to refactor using that.

File: examples/GroceryExample/readme.md (hunk 2 above)
Comment (victlu) [reply examples/GroceryExample/readme.md:16]: We have new APIs soon.  Let's address it then.

File: examples/GroceryExample/readme.md (hunk 3 above)
Comment (victlu) [reply examples/GroceryExample/readme.md:25]: We have new APIs soon.  Let's address it then.

File: examples/GroceryExample/GroceryStore.cs (hunk 6 above)
Comment (victlu) [reply examples/GroceryExample/GroceryStore.cs:93]: We have new APIs soon.  Let's address it then.

**Comment by victlu** [comment]
Ping to see if we should approve this?

**Comment by reyang** [comment]
> Ping to see if we should approve this?

I cannot approve this PR. I think the goal is to experiment + discuss this during the metrics API/SDK design, rather than ship code to the metrics branch at this phase.

**Pull request closed without merging**